File: 1930-unique-length-3-palindromic-subsequences/1930-unique-length-3-palindromic-subsequences.c
#define MAX_CHAR 26 // 알파벳 소문자 개수 -> 대문자 포함이면 52
int countUniqueCharsBetweenIndexes(char *s, int start, int end) {
    int hash_table[MAX_CHAR] = {0};
    for (int i = start + 1; i < end; i++) {
        if (s[i] >= 'a' && s[i] <= 'z') {
            hash_table[s[i] - 'a'] = 1;
        }
    }

    int count = 0;
    for (int i = 0; i < MAX_CHAR; i++) {
        count += hash_table[i];
    }
    return count;
}

int countPalindromicSubsequence(char* s) {
    int firstIndex[MAX_CHAR], lastIndex[MAX_CHAR];
    memset(firstIndex, -1, sizeof(firstIndex));
    memset(lastIndex, -1, sizeof(lastIndex));

    int len = strlen(s);
    for (int i = 0; i < len; i++) {
        int index = s[i] - 'a';
        if (firstIndex[index] == -1) firstIndex[index] = i;
        lastIndex[index] = i;
    }

    int count = 0;
    for (int i = 0; i < MAX_CHAR; i++) {
        if (firstIndex[i] != -1 && lastIndex[i] != -1 && firstIndex[i] != lastIndex[i]) {
            count += countUniqueCharsBetweenIndexes(s, firstIndex[i], lastIndex[i]);
        }
    }
    return count;
}

File: 1930-unique-length-3-palindromic-subsequences/1930-unique-length-3-palindromic-subsequences.py
class Solution:
    def countPalindromicSubsequence(self, s: str) -> int:
        def count_unique_chars_between_indexes(s, start, end):
            """주어진 인덱스 사이의 유니크한 문자 수를 계산합니다."""
            return len(set(s[start+1:end]))
        first_index = [-1] * 26
        last_index = [-1] * 26

        # 각 문자에 대해 첫 등장 위치와 마지막 등장 위치를 기록합니다.
        for i, char in enumerate(s):
            index = ord(char) - ord('a')
            if first_index[index] == -1:
                first_index[index] = i
            last_index[index] = i

        count = 0
        # 각 문자를 중심으로 가능한 팰린드롬 서브시퀀스 수를 계산합니다.
        for i in range(26):
            if first_index[i] != -1 and last_index[i] != -1 and first_index[i] != last_index[i]:
                count += count_unique_chars_between_indexes(s, first_index[i], last_index[i])

        return count